### ISSUE TITLE
Idea: Immutable Jenkins config and config as code

### ISSUE THREAD

**Issue by icereed** [jenkins-x/jenkins-x-image#34]
[DRAFT: This idea is not yet very nicely written down]
Hello together,

at my company we introduced a similar idea as Jenkins X. As these concepts are really close together, I would like to start contributing back ideas and feedback we got.

Two ideas:
# The config of Jenkins should be configurable via ConfigMap
# The Jenkins should only use the config given by the ConfigMap and reset any manual changes on restart

What is the benefit?
A fresh Jenkins installation is nice and clean. Over time, of course, as you get plugin updates, as people start messing around with the settings, the state of your Jenkins diverges from your infrastructure definition in your code. When we have everything defined in a ConfigMap (or more secure as secret) we are able to mitigate this issue. We would be free to change the image, plugins etc. as long as we keep the config working and consistent.

The ultimate goal you want to have, when you CICD platform is business critical, that you have a very easy recreation of the platform with all settings and plugins you desire. This is according to the Phoenix server ( https://martinfowler.com/bliki/PhoenixServer.html ) best practice.

We already achieved this by creating an ConfigMap which gets consumed by an init groovy script. Also wrote a very tiny framework to reduce code duplication per configuration item in the ConfigMap.

**Comment by icereed** [comment]
@rawlingsj That's pretty awesome! Could you elaborate how this is tested?

**Comment by rawlingsj** [comment]
Sure, we use Jenkins X to build and release Jenkins X. We have BDD tests https://github.com/jenkins-x/bdd-jx that run on each PR of https://github.com/jenkins-x/jx and then also before we publish a Jenkins X release which is done on PRs for this repo https://github.com/jenkins-x/cloud-environments.

The BDD tests build and start Jenkins then create our quickstarts ensuring they build test deploy and promote using their own pipelines.

**Comment by icereed** [comment]
Little addition: An idea came up to introduce a CRD with this in combination with the plugin.
https://github.com/jenkinsci/configuration-as-code-plugin/issues/272

**Comment by jstrachan** [comment]
yeah - we're hoping to migrate to use the configuration-as-code plugin ASAP - its already included as a plugin, we've just not swapped out all the `config.xml` files from the `ConfigMap` to the nice simple configuration as code yaml. 

The really awesome part of the configuration-as-code yaml is we can just use it as a big blob of yaml inside the helm `values.yaml` so that anyone can override, via helm, any parts of the configuration when they install Jenkins X

**Comment by icereed** [comment]
That was exactly what I was hoping! I will start to play around with the plugin to gain some knowledge. Maybe I'm fast enough to contribute. :)
